arbnode/batch_poster.go

```diff
@@ -882,37 +887,42 @@ type batchSegments struct {
 	lastCompressedSize    int
 	trailingHeaders       int // how many trailing segments are headers
 	isDone                bool
-	useCustomDA           bool // whether to use the DACertificate header flag
 }
 
 type buildingBatch struct {
 	segments           *batchSegments
 	startMsgCount      arbutil.MessageIndex
 	msgCount           arbutil.MessageIndex
 	haveUsefulMessage  bool
-	use4844            bool
-	useCustomDA        bool
+	use4844            bool // Are we ACTUALLY using 4844 for THIS batch?
+	wouldUse4844       bool // WOULD we use 4844 if we fell back to EthDA?
 	muxBackend         *simulatedMuxBackend
 	firstDelayedMsg    *arbostypes.MessageWithMetadata
 	firstNonDelayedMsg *arbostypes.MessageWithMetadata
 	firstUsefulMsg     *arbostypes.MessageWithMetadata
 }
 
-func (b *BatchPoster) newBatchSegments(ctx context.Context, firstDelayed uint64, use4844 bool, useCustomDA bool) (*batchSegments, error) {
-	maxSize := b.config().MaxSize
+func (b *BatchPoster) newBatchSegments(ctx context.Context, firstDelayed uint64, use4844 bool, usingAltDA bool) (*batchSegments, error) {
+	config := b.config()
+	maxSize := config.MaxSize
+
 	if use4844 {
-		if b.config().Max4844BatchSize != 0 {
-			maxSize = b.config().Max4844BatchSize
+		// Building 4844 blobs for EthDA
+		if config.Max4844BatchSize != 0 {
+			maxSize = config.Max4844BatchSize
 		} else {
 			maxBlobGasPerBlock, err := b.parentChain.MaxBlobGasPerBlock(ctx, nil)
 			if err != nil {
 				return nil, err
 			}
-			// Try to fill 3 blobs per batch
+			// Try to fill under half of the parent chain's max blobs.
 			// #nosec G115
 			maxSize = blobs.BlobEncodableData*(int(maxBlobGasPerBlock)/params.BlobTxBlobGasPerBlob)/2 - 2000
```

Comment:
this formula is pretty suspicious... as for my intuition, it should be like:
```golang
maxSize = sizeInASingleBlob * (numberOfBlobs / 2) - overhead
```

where:
 - taking "under half of the parent chain's max blobs", i.e. `numberOfBlobs / 2` is done **before** multiplying by single blob capacity (otherwise we might get an incorrect result)
 - we have an assumption that `int(maxBlobGasPerBlock)/params.BlobTxBlobGasPerBlob` is exactly the number of blobs available
 - we have an assumption that the overhead is used only in a single blob, not every one

Reply:
I'm not going to address this on the custom-da branch but we should follow up on it. There are a couple things:
1. We can make the formula more straight forward if we also use the target https://github.com/ethereum/go-ethereum/pull/32991
2.  Even if we wanted to submit more than 6, we can't https://eips.ethereum.org/EIPS/eip-7594
> Additionally, a limit of 6 blobs per transaction is introduced. Clients MUST enforce this limit when validating blob transactions at submission time, when received from the network, and during block production and processing.